### PR TITLE
Embedding Parameter Values set in Tokens work properly when `nil` or empty list

### LINKED ISSUE #40292: Static embedding - setting empty value for a locked filter prevents linked filters to be populated
### Describe the bug

In a dashboard shared with static embedding when an empty value is set for a locked filter, no filter values will be returned for any linked filters that depend on the locked filter.

### To Reproduce

1. Create a GUI question: Sample Database, select Products.
2. Save, and add it to a dashboard.
3. Set up filters for `Category` and `Title`. Link the `Title` filter to `Category`, so that only relevant titles are returned in the drop down.
4. Set up the dashboard for static embedding:
  * set `Category` parameter to Locked
  * leave `Title` as Editable

In the embedding application, set the `category` parameter in the JWT token to the following values, and load the page:
6. sending ({...,  "params":{"category":["Widget"] }} ) --> works ok (viz displays, Title dropdown has the proper values)
7. do not send the parameter --> works ok (results in a visualization error, API response: You must specify a value for :category in the JWT)
8(!) sending ({...,  "params":{"category":null}} ) --> viz displays, but the Title dropdown has no values
9(!) sending ({...,  "params":{"category":[] }} ) --> viz displays, but the Title dropdown has no values

### Expected behavior

In steps 8 and 9, the Title dropdown should be populated.

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.22+7",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.22",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.22+7",
    "os.name": "Linux",
    "os.version": "5.10.210-201.852.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "bigquery-cloud-sdk",
      "snowflake",
      "postgres",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "14.10"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-03-14",
      "tag": "v1.49.0",
      "hash": "46c668b"
    },
    "settings": {
      "report-timezone": "America/Los_Angeles"
    }
  }
}
```

### Severity

P1 - affects customers and I am not aware of a workaround.

### Additional context

Customer reported a similar error with v1.48.7, I used v1.49.0 to reproduce.

### PR DESCRIPTION
**Pull request by adam-james-v**
Fixes: #40292
[Slack convo](https://metaboat.slack.com/archives/C063Q3F1HPF/p1721388755876919?thread_ts=1719935985.293099&cid=C063Q3F1HPF)

If a locked parameter's value is set to `[]` in the embedding JWT, we want to treat this as 'no value has been set' on that filter, but we should still run the query as normal. Likewise, any linked filters should still properly derive their values from the unfiltered values of the filter to which they are linked.

When you're using a filter in the App, it makes intuitive sense that when you've not selected anything from a filter's
drop down list (or entered any text, etc.), then the filter is effectively not applied.

For token params, we should behave similarly; `[]` should be accepted, and signal that the user intends
for that filter to basically NOT apply. But, `nil` should behave as if the parameter has not been set at all, in which case we should expect the endpoint to return "You must specify a value for [parameter]".

This also means we should not pass the constraint to the linked filters, since it errors, but even if it worked, it
would return 0 results, making it impossible to properly apply the linked filter at all.

So, this change causes empty lists to be treated as nil, AND will remove any parameters with `nil` as a value before
being passed to the query processor, which results in the proper behaviour of not having constraints on the linked
filter, and having the 'parent' filter not filtering anything.

Notice `category 'null'`
image

Linked Filter correctly returns all values:
image

### CODE REVIEW

**Review by tsmacdonald — APPROVED** [review]
Looks reasonable

**Comment by tsmacdonald** [comment]
Test failure looks legit though: https://github.com/metabase/metabase/runs/27631472847

**Review by bshepherdson — APPROVED** [review]

**Review by deniskaber — COMMENTED** [review]
Hey @adam-james-v ! 
Not sure it works as expected:

https://github.com/user-attachments/assets/ef7356ac-bdf0-4b44-8474-82074d31028b

File: e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js [hunk 1]
```diff
@@ -559,12 +559,25 @@ describe("scenarios > embedding > dashboard parameters with defaults", () => {
       });
     });
 
-    visitIframe();
-    // The ID default (1 and 2) should apply, because it is disabled.
-    // The Name default ('Lina Heaney') should not apply, because the Name param is editable and unset
-    // The Source default ('Facebook') should not apply because the param is locked but the value is unset
-    // If either the Name or Source default applied the result would be 0.
-    cy.findByTestId("scalar-value").invoke("text").should("eq", "2");
+    cy.get("@dashboardId").then(dashboardId => {
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:562]: Well, with the code change in this PR, we should now expect when source is NOT specified in the token, that the query won't run and will warn 'you need to specify a value for source'.

**Comment by adam-james-v** [comment]
> Hey @adam-james-v ! Not sure it works as expected:
> Screen.Recording.2024-07-23.at.00.21.46.mov

Thanks for taking a look, @deniskaber.

What you're seeing is indeed the correct behavior now. 
Please see this thread where we concluded: https://metaboat.slack.com/archives/C063Q3F1HPF/p1721388755876919?thread_ts=1719935985.293099&cid=C063Q3F1HPF

Basically what you're seeing is the result of now treating `null` the same as a key not existing in the JWT at all. 
In your example scenario, if you scroll down in your preview settings, you should be able to set a value for the locked parameter 'category'. Once you do that, you should see the preview working as expected.

File: e2e/test/scenarios/embedding/embedding-native.cy.spec.js [hunk 2]
```diff
@@ -81,62 +82,61 @@ describe("scenarios > embedding > native questions", () => {
...
+        const payload = {
+          resource: { question: questionId },
+          params: { total: [] },
+        };
 
-      // eslint-disable-next-line no-unscoped-text-selectors -- deprecated usage
-      cy.contains("Organic");
-      // eslint-disable-next-line no-unscoped-text-selectors -- deprecated usage
-      cy.contains("Twitter").should("not.exist");
+        visitEmbeddedPage(payload);
```
Comment (WiNloSt) [inline e2e/test/scenarios/embedding/embedding-native.cy.spec.js:91]: This is a good change, but we shouldn't nest all the assertions unnecessarily.

Once we finished using `questionId` we should close the scope, so all the assertions should be left as is.

```ts
      cy.get("@questionId").then(questionId => {
        const payload = {
          resource: { question: questionId },
          params: { total: [] },
        };

        visitEmbeddedPage(payload);
      }) // close the block here.
```

File: e2e/test/scenarios/embedding/embedding-native.cy.spec.js [hunk 3]
```diff
@@ -337,15 +338,22 @@ describe("scenarios > embedding > native questions with default parameters", ()
...
+
+      visitEmbeddedPage(payload);
+
+      // Remove default filter value
+      clearFilterWidget();
+      // The ID default (1, 2) should apply, because it is disabled.
+      // The Name default ('Lina Heaney') should not apply, because the Name param is editable and empty
+      // The Source default ('Facebook') should not apply because the param is locked but the value is unset
+      // If either the Name or Source default applied the result would be 0.
+      cy.findByTestId("scalar-value").invoke("text").should("eq", "2");
```
Comment (WiNloSt) [inline e2e/test/scenarios/embedding/embedding-native.cy.spec.js:355]: Can you add a test that asserts a similar error message in the dashboard `There was a problem displaying this card`, but for question?

image

**Review by WiNloSt — APPROVED** [review]
I haven't reviewed the BE code, but the test behavior looks good. Can you add a test for question embed that asserts the error message when we don't pass `null` to a locked param? We already have a similar test for dashboard, but not for questions.

**Comment by WiNloSt** [comment]
> > Hey @adam-james-v ! Not sure it works as expected:
> > Screen.Recording.2024-07-23.at.00.21.46.mov
> 
> Thanks for taking a look, @deniskaber.
> 
> What you're seeing is indeed the correct behavior now. Please see this thread where we concluded: https://metaboat.slack.com/archives/C063Q3F1HPF/p1721388755876919?thread_ts=1719935985.293099&cid=C063Q3F1HPF
> 
> Basically what you're seeing is the result of now treating `null` the same as a key not existing in the JWT at all. In your example scenario, if you scroll down in your preview settings, you should be able to set a value for the locked parameter 'category'. Once you do that, you should see the preview working as expected.

@adam-james-v I think Denis was confused because the PR description isn't updated to our latest discussion on Slack, I have also added a link to the discussion in the PR description for easier access.

**Pull request merged**